source/board/microbitv2/nrf52820/power.c

```diff
@@ -19,19 +19,190 @@
  * limitations under the License.
  */
 
+#include "nrf.h"
+
+#include "gpio.h"
+#include "gpio_extra.h"
+#include "main_interface.h"
+#include "pwr_mon.h"
+#include "IO_Config.h"
+#include "DAP_config.h"
+#include "uart.h"
+#include "i2c.h"
+#include "rl_usb.h"
+
 #include "power.h"
 
+
+static void power_before(bool systemoff);
+static void power_after();
+static void power_systemoff();
+static void power_wfi();
+
+void POWER_CLOCK_IRQHandler(void);
+void GPIOTE_IRQHandler(void);
+
+
+extern volatile uint8_t wake_from_reset;
+extern volatile uint8_t wake_from_usb;
+extern volatile bool usb_pc_connected;
+extern main_usb_connect_t usb_state;
+extern power_source_t power_source;
+
+
+uint8_t  power_in_wfi;
+uint8_t  power_gpiote_enabled;
+uint32_t power_gpiote_intenset;
+
+
 void power_init()
 {
-    // TODO
+    power_in_wfi = 0;
+
+    gpio_cfg_input(GPIO_REG(RESET_BUTTON), GPIO_IDX(RESET_BUTTON), RESET_BUTTON_PULL);
+
+    // Enable NRF_POWER interrupt for USB removed/detected
+    NRF_POWER->INTENSET = POWER_INTENSET_USBREMOVED_Msk | POWER_INTENSET_USBDETECTED_Msk;
+    NVIC_EnableIRQ(POWER_CLOCK_IRQn);
 }
 
 void power_down()
 {
-    // TODO
+    power_systemoff();
 }
 
 void power_sleep()
 {
-    // TODO
+    power_wfi();
+}
+
+
+static void power_before(bool systemoff)
+{
+    // TODO - wait for USB transmit to computer to finish?
+    /* Wait for debug console output finished. */
+    // while (!(LPUART_STAT_TC_MASK & UART->STAT))
+    // {
+    // }
+
+    uart_uninitialize();
+
+    // TODO - I think maybe uart_uninitialize does this in USART_Uninitialize
```

Comment:
If you can confirm that then we can just add a comment indicating that, otherwise we should keep the pin uninit as well.